src/lib/work_selector/offset.ml

```diff
@@ -0,0 +1,87 @@
+open Core_kernel
+
+(** The work selection method implemented here is the following. We distinguish
+two cases:
+    1. The work pool is updated: a random [offset] is generated. The selected
+    work is the [offset]-th one of the expensive work list.
+    2. As long as the pool stays the same, the works are selected sequencially
+    from the offset. If the end of the work list is reached, the next works are
+    selected sequencially from the begining of the list. *)
+module Make
+    (Inputs : Intf.Inputs_intf)
+    (Lib : Intf.Lib_intf with module Inputs := Inputs) =
+struct
+  (* TODO randomness should be different for each worker, at each execution *)
+  module Offset = struct
+    (** The initialisation for the first offset can be 0; as the first
+   [previous_length] 0, any non-empty list will trigger an update of the offset. *)
+    let offset = ref 0
+
+    (** The previous length of expensive works [work] got through *)
+    let prev_length = ref 0
+
+    let get_offset () = !offset
+
+    (** This function maintains [prev_length] up do date, and if a Snark pool
+        update is detected, a new [offset] is generated.
+        When the pool is not updated, we consider that its length should
+        diminish by one at each call of [work]. If we don't observe that
+        behavior, the pool is considered updated and a new [offset] is 
+        generated randomly, uniformly chosen between 0 and [new_length]. *)
+    let update ~new_length =
+      let () =
+        if new_length = !prev_length - 1 then (
+          Printf.printf "\n  no offset update\n" ;
+          () )
+        else
+          let () =
+            let n = Random.int new_length in
+            offset := n ;
+            Printf.printf "\n  offset update to %d, new_length was %d\n" n
+              new_length
+          in
+          ()
+      in
+      prev_length := new_length
+
+    (** Because of the [offset] being constant and the [expensive_work] list
+    reducing in size, a case where [offset] ends up greater than the list
+    length may happen. In that case, we go back to the beginning of the list,
+    by resetting [offset] to 0.
+  /!\ fails if [expensive_work] is empty! *)
+    let get_nth expensive_work =
+      try
+        let res = List.nth_exn expensive_work !offset in
+        Printf.printf "\n" ; res
+      with _ ->
+        (* Raising an error here means that offset is now too large for the
+           list -> back to the beginning with offset at 0 *)
+        offset := 0 ;
+        Printf.printf "  offset reset to 0\n" ;
+        List.nth_exn expensive_work !offset
+  end
+
+  let work ~snark_pool ~fee ~logger (state : Lib.State.t) =
+    Lib.State.remove_old_assignments state ~logger ;
+    let unseen_jobs = Lib.State.all_unseen_works state in
+    match Lib.get_expensive_work ~snark_pool ~fee unseen_jobs with
+    | [] ->
+        None
+    | expensive_work ->
```

Comment:
I think we can achieve the intended result with:

```ocaml
| expensive_work ->
  let n = List.length expensive_work
  if offset >= n then 
    offset := Random.int n
  let x = List.nth_exn expensive !offset in (* shouldn't oob *)
  Lib.State.set state x ; 
  Some x
```

This way we process sequentially assuming the size change is one, similiar to the sequential solution.

Reply:
I suppose offset would have to initialize to the maximum size of the snark pool, or even int max